### PR TITLE
Improve `getsentinfo` accuracy

### LINKED ISSUE #2055: `getsentinfo` returns an indeterminate response `[]`
`getsentinfo` may return `[]` because it doesn't know anything about the given `--paymentHash` or `--id`, which means a payment wasn't performed or attempted, and is not pending.

But it may also return `[]` for a payment that is in process of being attempted, for some seconds or fractions of second before it is actually attempted, which has the opposite meaning.

For services that send payouts automatically on behalf of users and then check these payment statuses automatically this can lead to terrible loss of funds.

### PR DESCRIPTION
**Pull request by t-bast**
When using MPP, we previously inserted an entry in the payments DB only after path-finding completed. Meanwhile, `getsentinfo` would return an empty array, so users cannot know that the payment is being attempted. This could lead to users eagerly retrying and double-paying.

We now keep track of pending payments in the `PaymentInitiator`. When the payment doesn't seem to be pending based on the DB alone, we also check the `PaymentInitiator`.

However, this doesn't entirely fix the risk of double-paying, which cannot be completely fixed from inside eclair. External applications that make payments on behalf of users must use the `externalId` and check for double payments themselves before calling eclair if they want to ensure that there's no risk of double payments.

Each commit is independent, as I took this opportunity to do a bit of clean up.

Replaces #2126 to fix #2055

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala [hunk 1]
```diff
@@ -60,11 +60,13 @@ class PaymentInitiator(nodeParams: NodeParams, outgoingPaymentFactory: PaymentIn
           sender() ! PaymentFailed(paymentId, r.paymentHash, LocalFailure(r.recipientAmount, Nil, PaymentSecretMissing) :: Nil)
         case Some(paymentSecret) if r.paymentRequest.features.allowMultiPart && nodeParams.features.hasFeature(BasicMultiPartPayment) =>
           val fsm = outgoingPaymentFactory.spawnOutgoingMultiPartPayment(context, paymentCfg)
-          fsm ! SendMultiPartPayment(sender(), paymentSecret, r.recipientNodeId, r.recipientAmount, finalExpiry, r.maxAttempts, r.paymentRequest.paymentMetadata, r.assistedRoutes, r.routeParams, userCustomTlvs = r.userCustomTlvs)
+          fsm ! SendMultiPartPayment(self, paymentSecret, r.recipientNodeId, r.recipientAmount, finalExpiry, r.maxAttempts, r.paymentRequest.paymentMetadata, r.assistedRoutes, r.routeParams, userCustomTlvs = r.userCustomTlvs)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala:63]: An alternative to putting `replyTo = self` would have been to preserve the original sender and use a `watch`. In theory there would be a tiny performance gain, but I suppose you have favored consistency between trampoline and non-trampoline payments.

File: eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala [hunk 2]
```diff
@@ -375,11 +375,30 @@ class EclairImpl(appKit: Kit) extends Eclair with Logging {
     }
   }
 
-  override def sentInfo(id: Either[UUID, ByteVector32])(implicit timeout: Timeout): Future[Seq[OutgoingPayment]] = Future {
-    id match {
+  override def sentInfo(id: Either[UUID, ByteVector32])(implicit timeout: Timeout): Future[Seq[OutgoingPayment]] = {
+    val outgoingDbPayments = id match {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala:379]: The database call should be done within a `Future`, like before. It forces you to compose futures which is not very elegant but I think it's really important.

**Review by pm47 — COMMENTED** [review]
I like this version better. Given how far you went in the implementation I think you too?

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala:63]: I've always wanted to get rid of setting `replyTo` to `sender`, it was the simple thing to do but it introduced the annoying side effect that senders received the `PreimageReceived` event when using MPP, whereas they were only interested in the `PaymentSent` event. This was a good opportunity to fix that (and make trampoline and non-trampoline more consistent as well)!

File: eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala (hunk 2 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala]: Good catch, done in https://github.com/ACINQ/eclair/pull/2142/commits/bf8a61a7937712e49b41b2c97494196843c4e60f

**Comment by t-bast** [comment]
> I like this version better. Given how far you went in the implementation I think you too?

100% agree, this is much better, I don't have the reflex of putting more logic inside the API handlers themselves, but it makes a lot of sense in that case.

**Review by pm47 — APPROVED** [review]

**Comment by fiatjaf** [comment]
Does this mean there will never again be an empty `[]` if a payment is being tried? What will it return if the payment is being tried but not written to the DB yet?

What is this risk of double-payments you mention? Paying the same invoice twice?

**Comment by t-bast** [comment]
> Does this mean there will never again be an empty [] if a payment is being tried? 

Yes, there will always be something returned if you use a `pay` command then call `getsentinfo` on the id it returns.

> What will it return if the payment is being tried but not written to the DB yet?

It will return the same structure as it did if the payment was pending, but without route details (since no htlcs have been sent yet) and with `id == parentId`.

> What is this risk of double-payments you mention? Paying the same invoice twice?

Actually this completely solves the issue you had.
The only potential race condition that remains and cannot be fixed from within eclair would be:

- you call `payinvoice`
- you call `getsentinfo --paymentHash=<invoice_payment_hash>` before the previous call returned
- then it's not guaranteed that you will see a result

But as long as you wait for `payinvoice` (or other `pay` APIs) to return you the ID, eclair should now always return something in `getsentinfo`. I believe that completely fixes the issue you had.

**Pull request merged**